### PR TITLE
feat: support get datasource status

### PR DESCRIPTION
**Pull request by himself65**
We need a way to display the status of each data source

Especially in the cloud data source, there will have network pending and server error. There must have a way to block the UI

### CODE REVIEW

File: packages/y-indexeddb/src/provider.ts [hunk 1]
```diff
@@ -29,12 +30,24 @@ const createDatasource = ({
...
+    callbackSet.forEach(cb => cb());
+  };
+
   const dbPromise = openDB<BlockSuiteBinaryDB>(dbName, dbVersion, {
     upgrade: upgradeDB,
   });
   const adapter = {
     queryDocState: async (guid, options) => {
       try {
+        changeStatus({
```
Comment (pengx17) [inline packages/y-indexeddb/src/provider.ts:48]: I am not sure but it seems the way setting status may has race condition issues

File: packages/y-provider/src/types.ts [hunk 2]
```diff
@@ -1,3 +1,18 @@
+export type Status =
```
Comment (himself65) [inline packages/y-provider/src/types.ts:1]: Can we use `xstate` to handle this?

File: packages/y-indexeddb/src/provider.ts (hunk 1 above)
Comment (himself65) [reply packages/y-indexeddb/src/provider.ts]: I think you are right. we need a mutex lock here

Comment (himself65) [reply packages/y-indexeddb/src/provider.ts]: After I tried, I think it's not necessary to add a mutex here

Comment (himself65) [reply packages/y-indexeddb/src/provider.ts]: I simulate a stack, it works great

File: packages/y-provider/src/types.ts [hunk 3]
```diff
@@ -1,4 +1,24 @@
...
+  | {
+      type: 'syncing';
+    }
+  | {
+      type: 'synced';
+    }
+  | {
+      type: 'error';
+      error: Error;
+    };
```
[the remark this answers is not in the document]
Comment (himself65) [reply packages/y-provider/src/types.ts:14]: I think the `Status` here is just a signal for the UI. In the design part, we don't have such demand on display the sync details to the user

**Pull request merged**